### ISSUE TITLE
Sorting of `@psalm-return` intersection type may lead to detecting `mixed` instead of correct type

### ISSUE THREAD

**Issue by Ocramius** [vimeo/psalm#1677]
I am very sorry for reporting this one without a shorter reproduce scenario, but it seems to be a combination of sorting, inheritance priority, and class detection. I'll try putting the relevant code in here, but the issue is extracted from following code: https://github.com/Ocramius/ProxyManager/pull/465/commits/5357b0358d31b65c8115c93e8fa616035bff1e6e

```php
<?php

class MyProxiedClass
{
    /** @return string */
    public function sayHello()
    {
        return 'Hello!';
    }
}

/**
 * @psalm-template DecoratedClassName of object
 */
interface ProxyInterface
{
}

interface LazyLoadingInterface extends ProxyInterface
{
    public function initializeProxy() : bool;
}

/**
 * @psalm-template WrappedValueHolderType of object
 */
interface ValueHolderInterface extends ProxyInterface
{
    /**
     * @psalm-return WrappedValueHolderType|null
     */
    public function getWrappedValueHolderValue() : ?object;
}

interface VirtualProxyInterface extends LazyLoadingInterface, ValueHolderInterface
{
}


class LazyLoadingValueHolderFactory
{

    /**
     * @param array<string, mixed> $proxyOptions
     *
     * @psalm-template RealObjectType of object
     *
     * @psalm-param class-string<RealObjectType> $className
     *
     * @psalm-return RealObjectType&VirtualProxyInterface&ValueHolderInterface<RealObjectType>
     * @disabled-psalm-return RealObjectType&ValueHolderInterface<RealObjectType>&VirtualProxyInterface
     */
    public function createProxy(
        string $className,
        Closure $initializer,
        array $proxyOptions = []
    ) : VirtualProxyInterface {
        // ... snip ... 
    }
}

$valueHolder = (new LazyLoadingValueHolderFactory())
    ->createProxy(MyProxiedClass::class, function () {});

$valueHolder->initializeProxy();

$wrappedValue = $valueHolder->getWrappedValueHolderValue();

assert(null !== $wrappedValue);

echo $wrappedValue->sayHello();
```

https://psalm.dev/r/8a169678e8

`$wrappedValue` seems to be inferred to `mixed` when using this code, while using the second `@disabled-psalm-return`, it infers `MyProxiedClass`:

```diff
-     * @psalm-return RealObjectType&VirtualProxyInterface&ValueHolderInterface<RealObjectType>
-     * @disabled-psalm-return RealObjectType&ValueHolderInterface<RealObjectType>&VirtualProxyInterface
+     * @psalm-return RealObjectType&ValueHolderInterface<RealObjectType>&VirtualProxyInterface
```

**Comment by muglug** [comment]
I think this is caused by https://github.com/vimeo/psalm/issues/1686 - you have an unparameterised `VirtualProxyInterface` with a method `getWrappedValueHolderValue` that returns `object|null`, and that `object` supersedes the return type of the parameterised `ValueHolderInterface`

**Comment by muglug** [comment]
Fixed by 8b9de8b
